### DISCUSSION TITLE
When I use omp to read two h5 files concurrently, an exception occurs

### DISCUSSION THREAD

**Discussion by WanXinTao** [BlueBrain/HighFive#464] [Questions] [answered]
![image](https://user-images.githubusercontent.com/21972225/117577675-5600e800-b11d-11eb-84fa-7940b0d06bcc.png)

When I delete omp, serial execution is normal

I want to know what is the correct way for omp to read multiple h5 files in parallel, I think the introduction of highfive is to support parallel

**Comment by pramodk** [comment]
Other HighFive experts will help here but just want to point out that HighFive uses HDF5 library internally. So its worth reading HDF5 section about thread safety requirements: https://portal.hdfgroup.org/display/knowledge/Questions+about+thread-safety+and+concurrent+access

**Comment by ohm314** [comment]
HDF5 is not thread-safe. As @pramodk pointed out, HighFive uses the HDF5 library underneath, and hence using openmp (or other multithreading frameworks) to parallelize HighFive operations isn't supported.

**Comment by WanXinTao** [comment]
> HDF5 is not thread-safe. As @pramodk pointed out, HighFive uses the HDF5 library underneath, and hence using openmp (or other multithreading frameworks) to parallelize HighFive operations isn't supported.

If I compile a parallel HDF5, can highfive use the parallel function?

**Comment by ohm314** [comment, marked as answer]
Parallel HDF5 uses MPI to parallelize. While this won't give you thread-parallelism either, it is supported by HighFive. If you're looking to parallelize your HDF5 I/O, using distributed memory parallelism through HighFive with parallel HDF5 and MPI can be a viable choice, but that means that your application will have to be parallelized with MPI too.
